### PR TITLE
8271515: Integration of JEP 417: Vector API (Third Incubator)

### PR DESCRIPTION
**Pull request by PaulSandoz**
This PR improves the performance of vector operations that accept masks on architectures that support masking in hardware, specifically Intel AVX512 and ARM SVE.

On architectures that do not support masking in hardware the same technique as before is applied to most operations, specifically composition using blend.

Masked loads/stores are a special form of masked operation that require additional care to ensure out-of-bounds access throw exceptions. The range checking has not been fully optimized and will require further work.

No API enhancements were required and only a few additional tests were needed.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271515](https://bugs.openjdk.java.net/browse/JDK-8271515): Integration of JEP 417: Vector API (Third Incubator)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to f558b74ba4a3f60b4fb6cfd5f4f34fc25d0cff59
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**) ⚠️ Review applies to 101cd23a3c85ef5109510d6206681b2fd8f295d7
 * [Nick Gasson](https://openjdk.java.net/census#ngasson) (@nick-arm - **Reviewer**) ⚠️ Review applies to 571e6f39111c97ee3c9cb57765e6f26b31f9c279

### Contributors
 * Sandhya Viswanathan `<sviswanathan@openjdk.org>`
 * Jatin Bhateja `<jbhateja@openjdk.org>`
 * Ningsheng Jian `<njian@openjdk.org>`
 * Xiaohong Gong `<xgong@openjdk.org>`
 * Eric Liu `<eliu@openjdk.org>`
 * Jie Fu `<jiefu@openjdk.org>`
 * Vladimir Ivanov `<vlivanov@openjdk.org>`
 * John R Rose `<jrose@openjdk.org>`
 * Paul Sandoz `<psandoz@openjdk.org>`
 * Rado Smogura `<mail@smogura.eu>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5873/head:pull/5873` \
`$ git checkout pull/5873`

Update a local copy of the PR: \
`$ git checkout pull/5873` \
`$ git pull https://git.openjdk.java.net/jdk pull/5873/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5873`

View PR using the GUI difftool: \
`$ git pr show -t 5873`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5873.diff">https://git.openjdk.java.net/jdk/pull/5873.diff</a>

</details>

### CODE REVIEW

**Comment by PaulSandoz** [comment]
@njian there is a conflict with `macroAssembler_aarch64.cpp`:

```
--- a/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
@@@ -2580,8 -2495,8 +2571,13 @@@ void MacroAssembler::pop_call_clobbered
  }
  
  void MacroAssembler::push_CPU_state(bool save_vectors, bool use_sve,
++<<<<<<< HEAD
 +                                    int sve_vector_size_in_bytes, int total_predicate_in_bytes) {
 +  push(0x3fffffff, sp);         // integer registers except lr & sp
++=======
+                                     int sve_vector_size_in_bytes) {
+   push(RegSet::range(r0, r29), sp); // integer registers except lr & sp
++>>>>>>> master
    if (save_vectors && use_sve && sve_vector_size_in_bytes > 16) {
      sub(sp, sp, sve_vector_size_in_bytes * FloatRegisterImpl::number_of_registers);
      for (int i = 0; i < FloatRegisterImpl::number_of_registers; i++) {
```

I think the resolution is this:

```
@@ -2581,7 +2572,7 @@ void MacroAssembler::pop_call_clobbered_registers_except(RegSet exclude) {
 
 void MacroAssembler::push_CPU_state(bool save_vectors, bool use_sve,
                                     int sve_vector_size_in_bytes, int total_predicate_in_bytes) {
-  push(0x3fffffff, sp);         // integer registers except lr & sp
+  push(RegSet::range(r0, r29), sp); // integer registers except lr & sp
```

is that correct?

**Comment by PaulSandoz** [comment]
@njian there is a conflict with `macroAssembler_aarch64.cpp`:

```
--- a/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
@@@ -2580,8 -2495,8 +2571,13 @@@ void MacroAssembler::pop_call_clobbered
  }
  
  void MacroAssembler::push_CPU_state(bool save_vectors, bool use_sve,
++<<<<<<< HEAD
 +                                    int sve_vector_size_in_bytes, int total_predicate_in_bytes) {
 +  push(0x3fffffff, sp);         // integer registers except lr & sp
++=======
+                                     int sve_vector_size_in_bytes) {
+   push(RegSet::range(r0, r29), sp); // integer registers except lr & sp
++>>>>>>> master
    if (save_vectors && use_sve && sve_vector_size_in_bytes > 16) {
      sub(sp, sp, sve_vector_size_in_bytes * FloatRegisterImpl::number_of_registers);
      for (int i = 0; i < FloatRegisterImpl::number_of_registers; i++) {
```

I think the resolution is this:

```
@@ -2581,7 +2572,7 @@ void MacroAssembler::pop_call_clobbered_registers_except(RegSet exclude) {
 
 void MacroAssembler::push_CPU_state(bool save_vectors, bool use_sve,
                                     int sve_vector_size_in_bytes, int total_predicate_in_bytes) {
-  push(0x3fffffff, sp);         // integer registers except lr & sp
+  push(RegSet::range(r0, r29), sp); // integer registers except lr & sp
```

is that correct?

**Comment by nsjian** [comment]
> @njian there is a conflict with `macroAssembler_aarch64.cpp`:
> 
> ```
> --- a/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
> +++ b/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
> @@@ -2580,8 -2495,8 +2571,13 @@@ void MacroAssembler::pop_call_clobbered
>   }
>   
>   void MacroAssembler::push_CPU_state(bool save_vectors, bool use_sve,
> ++<<<<<<< HEAD
>  +                                    int sve_vector_size_in_bytes, int total_predicate_in_bytes) {
>  +  push(0x3fffffff, sp);         // integer registers except lr & sp
> ++=======
> +                                     int sve_vector_size_in_bytes) {
> +   push(RegSet::range(r0, r29), sp); // integer registers except lr & sp
> ++>>>>>>> master
>     if (save_vectors && use_sve && sve_vector_size_in_bytes > 16) {
>       sub(sp, sp, sve_vector_size_in_bytes * FloatRegisterImpl::number_of_registers);
>       for (int i = 0; i < FloatRegisterImpl::number_of_registers; i++) {
> ```
> 
> I think the resolution is this:
> 
> ```
> @@ -2581,7 +2572,7 @@ void MacroAssembler::pop_call_clobbered_registers_except(RegSet exclude) {
>  
>  void MacroAssembler::push_CPU_state(bool save_vectors, bool use_sve,
>                                      int sve_vector_size_in_bytes, int total_predicate_in_bytes) {
> -  push(0x3fffffff, sp);         // integer registers except lr & sp
> +  push(RegSet::range(r0, r29), sp); // integer registers except lr & sp
> ```
> 
> is that correct?

Yes, I think so.

**Comment by nsjian** [comment]
> C2 and x86 changes seems fine.

Could someone please help to review AArch64 changes in this patch? @theRealAph @adinn @nick-arm ?

The AArch64 changes include:
1. SVE scalable predicate register allocation support
2. SVE backend support for vector masking operations with predicate feature

**Comment by sviswa7** [comment]
The Java changes look good to me.

File: src/hotspot/cpu/aarch64/aarch64_sve_ad.m4 [hunk 1]
```diff
@@ -1664,369 +2001,391 @@ VSHIFT_COUNT(vshiftcntS, H,  8, T_SHORT)
...
+  match(Set dst (VectorCastD2X src));
+  effect(TEMP_DEF dst, TEMP tmp);
+  ins_cost(3 * SVE_COST);
+  format %{ "sve_vectorcast_d2x  $dst, $src\t# convert D to X vector (narrow)" %}
   ins_encode %{
-    __ sve_$3(as_FloatRegister($dst$$reg), __ $4, ptrue, as_FloatRegister($src$$reg), __ $4);
-    __ sve_$5(as_FloatRegister($dst$$reg), __ $6, as_FloatRegister($dst$$reg));
+    BasicType to_bt = Matcher::vector_element_basic_type(this);
+    Assembler::SIMD_RegVariant to_size = __ elemType_to_regVariant(to_bt);
+    __ sve_fcvtzs(as_FloatRegister($dst$$reg), __ D, ptrue, as_FloatRegister($src$$reg), __ D);
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_sve_ad.m4:2349]: Converting from double to long and then narrow to target types did not follow JLS. I will fix it. Thanks to @fg1417 for helping to find out this issue.

Comment (nsjian) [reply src/hotspot/cpu/aarch64/aarch64_sve_ad.m4]: > Converting from double to long and then narrow to target types did not follow JLS. I will fix it. Thanks to @fg1417 for helping to find out this issue.

Fixed in the new commit. Thanks to @PaulSandoz for integrating the fix!

Hi Nick @nick-arm ,

Could you please help to review the new commit, which fixes the same issue as JDK-8276151 for SVE? Thanks!

**Review by nick-arm — APPROVED** [review]

**Pull request closed without merging**